internal/timeline/timeline.go

```diff
@@ -0,0 +1,195 @@
+package timeline
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+const EPSILON TIME = 0.0001
+const DEFAULT_TRANSITION TIME = 0.5
+
+var debugTimeline bool = false
+var idSeed int64
+var idSeedMutex sync.Mutex
+
+type ON_STEP func(*TIME) ITimeline
+type ON_ACTIVE func(bool)
+
+type ITimeline interface {
+	GetTimeline() *Timeline
+	Step(time *TIME) ITimeline
+	SetActive(bool)
+}
+
+type Timeline struct {
+	ID           int64
+	active       bool
+	offset       TIME
+	speed        float64
+	fadeIn       *Interval
+	fadeOut      *Interval
+	freezingTime TIME
+	next         ITimeline
+	group        *TimelineGroup
+	onStep       ON_STEP
+	onActive     ON_ACTIVE
+}
+
+func (t *Timeline) Init() *Timeline {
+	idSeedMutex.Lock()
+	defer idSeedMutex.Unlock()
+	t.ID = idSeed
+	idSeed++
+	if debugTimeline {
+		log.Printf("Timeline %X", t.ID)
+	}
+	t.speed = 1.0
+	return t
+}
+
+func (t *Timeline) GetTimeline() *Timeline {
+	return t
+}
+
+func (t *Timeline) SetActive(on bool) {
+	if t.active == on {
+		return
+	}
+	if debugTimeline {
+		log.Println("SetActive", on, t.ID)
+	}
+	t.active = on
+	if t.onActive != nil {
+		t.onActive(on)
+	}
+}
+
+func (t *Timeline) Step(time *TIME) ITimeline {
+	if *time < 0.0 {
+		*time = 0.0
+	}
+
+	var running ITimeline = t
+	var LOOP_LIMIT int = 10000
```

Comment:
Please refer to the naming conventions of other variable in the project.

Reply:
LOOP_LIMIT will be replaced by loopLimit